src/bindings/python/PyConfig.cpp

```diff
@@ -1,1093 +1,601 @@
+                                                                       i);
+                return it.m_obj->getColorSpace(name);
+            });
+
+    py::class_<ActiveColorSpaceNameIterator>(cls, "ActiveColorSpaceNameIterator")
+        .def("__len__", [](ActiveColorSpaceNameIterator & it) 
+            { 
+                return it.m_obj->getNumColorSpaces(); 
+            })
+        .def("__getitem__", [](ActiveColorSpaceNameIterator & it, int i) 
+            { 
+                it.checkIndex(i, it.m_obj->getNumColorSpaces());
+                return it.m_obj->getColorSpaceNameByIndex(i);
+            })
+        .def("__iter__", [](ActiveColorSpaceNameIterator & it) -> ActiveColorSpaceNameIterator & 
+            { 
+                return it; 
+            })
+        .def("__next__", [](ActiveColorSpaceNameIterator & it)
+            {
+                int i = it.nextIndex(it.m_obj->getNumColorSpaces());
+                return it.m_obj->getColorSpaceNameByIndex(i);
+            });
+
+    py::class_<ActiveColorSpaceIterator>(cls, "ActiveColorSpaceIterator")
+        .def("__len__", [](ActiveColorSpaceIterator & it) 
+            { 
+                return it.m_obj->getNumColorSpaces(); 
+            })
+        .def("__getitem__", [](ActiveColorSpaceIterator & it, int i) 
+            { 
+                it.checkIndex(i, it.m_obj->getNumColorSpaces());
+                const char * name = it.m_obj->getColorSpaceNameByIndex(i);
+                return it.m_obj->getColorSpace(name);
+            })
+        .def("__iter__", [](ActiveColorSpaceIterator & it) -> ActiveColorSpaceIterator & 
+            { 
+                return it; 
+            })
+        .def("__next__", [](ActiveColorSpaceIterator & it)
+            {
+                int i = it.nextIndex(it.m_obj->getNumColorSpaces());
+                const char * name = it.m_obj->getColorSpaceNameByIndex(i);
+                return it.m_obj->getColorSpace(name);
+            });
+
+    py::class_<RoleNameIterator>(cls, "RoleNameIterator")
+        .def("__len__", [](RoleNameIterator & it) { return it.m_obj->getNumRoles(); })
+        .def("__getitem__", [](RoleNameIterator & it, int i) 
+            { 
+                it.checkIndex(i, it.m_obj->getNumRoles());
+                return it.m_obj->getRoleName(i);
+            })
+        .def("__iter__", [](RoleNameIterator & it) -> RoleNameIterator & { return it; })
+        .def("__next__", [](RoleNameIterator & it)
+            {
+                int i = it.nextIndex(it.m_obj->getNumRoles());
+                return it.m_obj->getRoleName(i);
+            });
+
+    py::class_<RoleColorSpaceIterator>(cls, "RoleColorSpaceIterator")
+        .def("__len__", [](RoleColorSpaceIterator & it) { return it.m_obj->getNumRoles(); })
+        .def("__getitem__", [](RoleColorSpaceIterator & it, int i) 
+            { 
+                it.checkIndex(i, it.m_obj->getNumRoles());
```

Comment:
I reviewed each of these methods in the C++ API. All of them with `it.checkIndex()` return nullptr instead of throwing, so this check throws when the index is out of range instead. The cases where an exception was already thrown, I omitted `it.checkIndex()` with a comment as to why. For example, in `PyGroupTransform.cpp:65`, in the `TransformIterator`, there is a comment reading `// GroupTransform provides index check with exception`. I agree that consistency in C++ would be nice eventually, at which point these checks could be removed. I know that has fairly wide implications for the API though.